### ISSUE TITLE
[microprofile] The /openapi throw exception `OpenAPIBuildException: java.lang.StackOverflowError` /FISH-48 /FISH-390

### ISSUE THREAD

**Issue by charleech** [payara/Payara#4818]
# Description #
----------

Just create the simple `JAXRS` service and deploy to the `Payara-Micro version 5.2020.3` and access to the `/openapi`. The console gives us the exception as the following: -

```
  OpenAPI document creation failed.
fish.payara.microprofile.openapi.api.OpenAPIBuildException: java.lang.StackOverflowError
        at fish.payara.microprofile.openapi.impl.OpenApiService$OpenApiMapping.buildDocument(OpenApiService.java:284)
        at fish.payara.microprofile.openapi.impl.OpenApiService$OpenApiMapping.access$200(OpenApiService.java:248)
        at fish.payara.microprofile.openapi.impl.OpenApiService.getDocument(OpenApiService.java:203)
        at fish.payara.microprofile.openapi.impl.rest.app.service.OpenApiResource.getResponse(OpenApiResource.java:80)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory.lambda$static$0(ResourceMethodInvocationHandlerFactory.java:52)
        at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:124)
        at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:167)
        at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$ResponseOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:176)
        at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:79)
        at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:469)
        at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:391)
        at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:80)
        at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:253)
        at org.glassfish.jersey.internal.Errors$1.call(Errors.java:248)
        at org.glassfish.jersey.internal.Errors$1.call(Errors.java:244)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:292)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:274)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:244)
        at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:265)
        at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:232)
        at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:680)
        at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:394)
        at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:346)
        at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:366)
        at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:319)
        at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:205)
        at org.apache.catalina.core.StandardWrapper.service(StandardWrapper.java:1636)
        at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:259)
        at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:161)
        at org.apache.catalina.core.StandardPipeline.doInvoke(StandardPipeline.java:757)
        at org.apache.catalina.core.StandardPipeline.invoke(StandardPipeline.java:577)
        at com.sun.enterprise.web.WebPipeline.invoke(WebPipeline.java:99)
        at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:158)
        at org.apache.catalina.connector.CoyoteAdapter.doService(CoyoteAdapter.java:371)
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:238)
        at com.sun.enterprise.v3.services.impl.ContainerMapper$HttpHandlerCallable.call(ContainerMapper.java:520)
        at com.sun.enterprise.v3.services.impl.ContainerMapper.service(ContainerMapper.java:217)
        at org.glassfish.grizzly.http.server.HttpHandler.runService(HttpHandler.java:182)
        at org.glassfish.grizzly.http.server.HttpHandler.doHandle(HttpHandler.java:156)
        at org.glassfish.grizzly.http.server.HttpServerFilter.handleRead(HttpServerFilter.java:218)
        at org.glassfish.grizzly.filterchain.ExecutorResolver$9.execute(ExecutorResolver.java:95)
        at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeFilter(DefaultFilterChain.java:260)
        at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeChainPart(DefaultFilterChain.java:177)
        at org.glassfish.grizzly.filterchain.DefaultFilterChain.execute(DefaultFilterChain.java:109)
        at org.glassfish.grizzly.filterchain.DefaultFilterChain.process(DefaultFilterChain.java:88)
        at org.glassfish.grizzly.ProcessorExecutor.execute(ProcessorExecutor.java:53)
        at org.glassfish.grizzly.nio.transport.TCPNIOTransport.fireIOEvent(TCPNIOTransport.java:524)
        at org.glassfish.grizzly.strategies.AbstractIOStrategy.fireIOEvent(AbstractIOStrategy.java:89)
        at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.run0(WorkerThreadIOStrategy.java:94)
        at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.access$100(WorkerThreadIOStrategy.java:33)
        at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy$WorkerThreadRunnable.run(WorkerThreadIOStrategy.java:114)
        at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.doWork(AbstractThreadPool.java:569)
        at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.run(AbstractThreadPool.java:549)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.StackOverflowError
        at org.glassfish.hk2.classmodel.reflect.impl.TypeProxy$1.iterator(TypeProxy.java:108)
        at fish.payara.microprofile.openapi.impl.model.util.AnnotationInfo.init(AnnotationInfo.java:269)
        at fish.payara.microprofile.openapi.impl.model.util.AnnotationInfo.init(AnnotationInfo.java:271)
        at fish.payara.microprofile.openapi.impl.model.util.AnnotationInfo.init(AnnotationInfo.java:271)
        at fish.payara.microprofile.openapi.impl.model.util.AnnotationInfo.init(AnnotationInfo.java:271)
        at fish.payara.microprofile.openapi.impl.model.util.AnnotationInfo.init(AnnotationInfo.java:271)
        at fish.payara.microprofile.openapi.impl.model.util.AnnotationInfo.init(AnnotationInfo.java:271)
```

## Expected Outcome

Should not have any error and display the `OpenAPI `response

## Current Outcome

The console show the exception and the response is empty.

## Steps to reproduce (Only for bug reports) 

1. Start application with `Payara-Micro`

```bash
java -jar payara-micro-5.2020.3.jar \
--deploy payara-evaluation-config-1.0.0-SNAPSHOT.war \
--contextroot / \
--addLibs ./libs
```

2. Access the `/openapi`

```
curl -v \
-H "Accept: application/json" \
-X GET \
http://localhost:8080/openapi
```

### Samples

The full maven project which reproduces this issue is at [charleech/payara-evaluation](https://github.com/charleech/payara-evaluation/tree/master/payara-evaluation-config)

## Context (Optional)

N/A

## Environment ##

- **Payara Version**: 5.2020.3
- **Edition**: Micro
- **JDK Version**: 
```
java version "1.8.0_181"
Java(TM) SE Runtime Environment (build 1.8.0_181-b13)
Java HotSpot(TM) 64-Bit Server VM (build 25.181-b13, mixed mode)
```
- **Operating System**: 

```
CentOS Linux release 7.7.1908 (Core)
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
```

- **Database**: N/A

**Comment by LarsBrenker** [comment]
I have the same problem. But only if i set "mp.openapi.scan.lib=true" in MP Config.

**Comment by jGauravGupta** [comment]
Hi @LarsBrenker,

Can you share the sample reproducer or simply the lib folder, we already fixed a similar issue in the HK2 class model parser.

**Comment by charleech** [comment]
Hi @jGauravGupta 

I'm not sure if my reproducer for https://github.com/payara/Payara/issues/3769#issuecomment-483564073 which is marked as fixed by #4768 is suitable or reflect the issue about jar scanning and this stack overflow error or not.

**Comment by LarsBrenker** [comment]
> 
> 
> Hi @LarsBrenker,
> 
> Can you share the sample reproducer or simply the lib folder, we already fixed a similar issue in the HK2 class model parser.

I created a minimal configuration and found out that the problem is related to the Log4J libs.

```
<dependency>
  <groupId>org.apache.logging.log4j</groupId>
  <artifactId>log4j-web</artifactId>
  <version>2.11.0</version>
</dependency>
        
<dependency>
  <groupId>org.apache.logging.log4j</groupId>
  <artifactId>log4j-slf4j-impl</artifactId>
  <version>2.11.0</version>
</dependency>
```

If I removed the libs, I have no problem generating the OpenApi files.

However, I also found out that the error occurs after starting Payara and calling /openapi for the first time. Do I generate the document again afterwards e.g. the document is displayed via the browser refresh button.
[SamplePom.txt](https://github.com/payara/Payara/files/5090745/SamplePom.txt)

**Comment by rdebusscher** [comment]
Hi Lars,

Thanks for the reproducer. I have created issue `FISH-390` based on your info.

Regards
Rudy

**Comment by rdebusscher** [comment]
Hi,

Both StackOverflowError will be solved in next community release.

Regards
Rudy

**Comment by poikilotherm** [comment]
Hi folks,
in IQSS/dataverse#5794 we experience the same stack overflow issue. @smillidge pointed out that the above mentioned log4j packages are causing those troubles.

Yet, in the Dataverse project those aren't used either as direct nor transient dependencies (verified via dependency tree and content of WAR). Is there something else that might cause troubles which hasn't been discovered yet?

**Comment by charleech** [comment]
The comments sorted by timeline are as the following: -

The https://github.com/payara/Payara/issues/4802#issuecomment-669830795 mentions that

> Looks like this issue occurs when the return type is Response.

The https://github.com/payara/Payara/pull/4832#issuecomment-673288939 mentions that

> #4818 SOF issue is fixed by the Patch HK2 v2.6.1.payara-p3.

And here https://github.com/payara/Payara/issues/4818#issuecomment-675490044
> I created a minimal configuration and found out that the problem is related to the Log4J libs.

For me at this issue, the reproducer at [charleech/payara-evaluation](https://github.com/charleech/payara-evaluation/tree/master/payara-evaluation-config) has its  dependency tree as the following: -

```bash
 $ mvn dependency:tree -Dscope=runtime
[INFO] --- maven-dependency-plugin:3.1.2:tree (default-cli) @ payara-evaluation-config ---
[INFO] Verbose not supported since maven-dependency-plugin 3.0
[INFO] app.github.charleech:payara-evaluation-config:war:1.0.0-SNAPSHOT
[INFO] +- app.github.charleech:payara-evaluation-base:jar:1.0.0-SNAPSHOT:compile
[INFO] +- org.slf4j:jcl-over-slf4j:jar:1.7.30:compile
[INFO] +- org.slf4j:jul-to-slf4j:jar:1.7.30:compile
[INFO] +- org.slf4j:log4j-over-slf4j:jar:1.7.30:compile
[INFO] +- org.slf4j:slf4j-api:jar:1.7.30:compile
[INFO] +- org.slf4j:slf4j-ext:jar:1.7.30:compile
[INFO] +- ch.qos.logback:logback-classic:jar:1.2.3:compile
[INFO] \- ch.qos.logback:logback-core:jar:1.2.3:compile
[INFO] ------------------------------------------------------------------------
```

As far as I understand, there are several root causes which produce this stack overflow issue and one of them is the `log4j` as _smillidge_ mentioning. 

Please correct me If I'm wrong. It would be nice, if we have more _reproducer_ which reflect and cover more use cases so that the next release may uses them for regression testing and ensuring this enhancement.

**Comment by LarsBrenker** [comment]
I don't know if the reproducer is still needed, but I created one for https://github.com/payara/Payara/issues/4818#issuecomment-675490044
https://github.com/LarsBrenker/pay4818

**Comment by fturizo** [comment]
Fixed as part of release *5.2020.5*

**Comment by LarsBrenker** [comment]
I looked at it again in 5.2020.7 and noticed that it now works. To close the subject properly I just wanted to say that the problem doesn't seem to exist anymore
